cpp/src/strings/search/findall_record.cu

```diff
@@ -0,0 +1,167 @@
+      auto const epos = d_str.byte_offset(end);    // to bytes
+
+      d_output[output_idx++] = string_index_pair{d_str.data() + spos, (epos - spos)};
+
+      begin = end + (begin == end);
+      end   = d_str.length();
+    }
+  }
+};
+
+}  // namespace
+
+//
+std::unique_ptr<column> findall_record(
+  strings_column_view const& strings,
+  std::string const& pattern,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  auto const strings_count = strings.size();
+  auto const d_strings     = column_device_view::create(strings.parent(), stream);
+
+  // compile regex into device object
+  auto const d_prog =
+    reprog_device::create(pattern, get_character_flags_table(), strings_count, stream);
+
+  // Create lists offsets column
+  auto offsets   = count_matches(*d_strings, *d_prog, stream, mr);
+  auto d_offsets = offsets->mutable_view().data<offset_type>();
+
+  // Compute null output rows
+  auto [null_mask, null_count] = cudf::detail::valid_if(
+    d_offsets, d_offsets + strings_count, [] __device__(auto v) { return v > 0; }, stream, mr);
+
+  auto const valid_count = strings_count - null_count;
+  // Return an empty lists column if there are no valid rows
+  if (valid_count == 0) {
+    return make_lists_column(0,
+                             make_empty_column(type_to_id<offset_type>()),
+                             make_empty_column(type_id::STRING),
+                             0,
+                             rmm::device_buffer{},
+                             stream,
+                             mr);
+  }
+
+  // Convert counts into offsets
+  thrust::exclusive_scan(
+    rmm::exec_policy(stream), d_offsets, d_offsets + strings_count + 1, d_offsets);
+
+  // Create indices vector with the total number of groups that will be extracted
+  auto total_matches = cudf::detail::get_value<size_type>(offsets->view(), strings_count, stream);
+
+  rmm::device_uvector<string_index_pair> indices(total_matches, stream);
+  auto d_indices = indices.data();
+  auto begin     = thrust::make_counting_iterator<size_type>(0);
+
+  // Build the string indices
+  auto const regex_insts = d_prog->insts_counts();
+  if (regex_insts <= RX_SMALL_INSTS) {
+    findall_fn<RX_STACK_SMALL> fn{*d_strings, *d_prog, d_offsets, d_indices};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, fn);
+  } else if (regex_insts <= RX_MEDIUM_INSTS) {
+    findall_fn<RX_STACK_MEDIUM> fn{*d_strings, *d_prog, d_offsets, d_indices};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, fn);
+  } else if (regex_insts <= RX_LARGE_INSTS) {
+    findall_fn<RX_STACK_LARGE> fn{*d_strings, *d_prog, d_offsets, d_indices};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, fn);
+  } else {
+    findall_fn<RX_STACK_ANY> fn{*d_strings, *d_prog, d_offsets, d_indices};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, fn);
+  }
+
```

Comment:
This dispatch pattern must be common throughout regex codes. Could we implement a dispatcher for this based on `floor(regex_insts)`? Feel free to push to future work, no need to clutter this PR even if you think it's a good idea, I just wanted to mention it while it's fresh.

Reply:
That is interesting. I will think on it for future work. Thanks.

Reply:
Created #10138 for this.